### ISSUE TITLE
Why is key_wise_distance selected with [topk]？

### ISSUE THREAD

**Issue by ADCarry0322** [KHU-AGI/Si-Blurry#1]
Hello,

I noticed an implementation detail in the similarity_loss computation when using the contrastive learning mode (self.use_contrastiv). The current code snippet is as follows:
`if self.use_contrastiv:
    key_wise_distance = 1 - F.cosine_similarity(self.key.unsqueeze(1), self.key, dim=-1)
    self.similarity_loss = -((key_wise_distance[topk] / mass[topk]).exp().mean() / ((distance / mass[topk]).exp().mean() + (key_wise_distance[topk] / mass[topk]).exp().mean()) + 1e-6).log()
`
From my understanding of the LCVPT loss as outlined in the original paper, the s_n (negative sample similarity score) formula requires us to compute the similarity (or distance) between all keys, not just the selected top-k keys. In the current implementation, however, it seems like key_wise_distance is indexed using [topk], which only considers a subset of keys (top-k). Shouldn't s_n be calculated over the entire set of key features instead of a subset?

**Comment by moonjunyyy** [comment]
Thank you for your interest in our work and for bringing this issue to our attention.

It appears that the implementation of the section you mentioned was designed to prevent the model from learning keys that were not selected at the time. Specifically, the key_wise_distance is a matrix of size key_size * key_size. Our intention was to select a subset of top-k elements to ensure that we only learn the distances from the selected keys to all other keys.

Since the selection size is fixed to 1, as mentioned in our limitations, this effectively computes the distance to all keys. However, we did not properly detach the negative pairs in this process, which unintentionally allowed them to be updated during training.

We appreciate your valuable feedback on our work and its implementation. We are actively conducting research based on this code and will ensure that this issue is reviewed, fixed, and validated promptly.

Thank you again for your contribution.

**Comment by moonjunyyy** [comment]
Thank you once again for your interest in our project and for bringing these points to our attention.

I apologize for the delayed response; I am currently traveling for a conference and have just now been able to review your comments.

Because of the newly discovered issue regarding negative sample updates, I have implemented a code update that should properly address this matter. However, the specific issue you pointed out is not the root cause of the problem we are currently experiencing, as I had hoped 😂.

Nonetheless, your feedback is invaluable, and I appreciate your efforts to help improve the project. If you have any further questions or encounter any other issues, please feel free to contact us via GitHub issues or email.
